### ISSUE TITLE
PDP11 autoconfig of RY (RX211) Unibus address/vector broken

### ISSUE THREAD

**Issue by agn453** [simh/simh#263]
While trying to read an RT-11 format RX02 floppy image under RSTS/E, I disabled the RX11 device (RX) and enabled a RX211 (RY) and notice that the address and vector allocated are floating ones and not 17777170/264 for the current version of SIMH. This has worked for me in the past (more than 5 years ago)

PDP-11 simulator V4.0-0 Beta        git commit id: aa5bc9e8
sim> set cpu 11/70
Disabling XQ
sim> show rx
RX      address=17777170-17777173, vector=264, 2 units
  RX0   256KB, not attached, write enabled
  RX1   256KB, not attached, write enabled
sim> set rx dis
sim> set ry ena
sim> sh ry
RY      address=17760270-17760273*, vector=340*, 2 units
  RY0   512KB, not attached, write enabled
        double density
  RY1   512KB, not attached, write enabled
        double density
sim> 

Under SIMH 3.8-1 I see the expected behaviour (RY at the correct address/vector) -

PDP-11 simulator V3.8-1
sim> set cpu 11/70
Disabling XQ
sim> show rx
RX, address=17777170-17777173*, vector=264, 2 units
  RX0, 256KB, not attached, write enabled
  RX1, 256KB, not attached, write enabled
sim> set rx dis
sim> set ry ena
sim> show ry
RY, address=17777170-17777173*, vector=264, 2 units
  RY0, 512KB, not attached, write enabled, double density
  RY1, 512KB, not attached, write enabled, double density
sim>

I tried manually setting the address and vector for RY, and this works - but then some other devices cannot be setup manually with autoconfig disabled (e.g. I can't set the interrupt vector on the DELUA (XU) and TS11 (TS) devices - no SET xx VECTOR=yyy command is available).

If I get a chance, I'll try and work out what's changed.

Tony

**Comment by markpizz** [comment]
Hi Tony,

MANY things have changed.  Previously many/most devices were statically assigned addresses.  This was OK when the set of devices assigned addresses in the floating address space was small and generally not very changeable.  When the set of floating devices got larger many things were moving all over the place and the static definitions for some devices started to break.  The solution was to pull out all of the static definitions and to assign all addresses and vectors via the autoconfigure mechanism.

You have encountered an instance where the current logic which walks through the autoconfigure table doesn't work.  Now that you've got me thinking about this, there are actually other examples which also demonstrate the same problem: Disable XQ and enable XQB.  XQB should end up with the static address that XQ would normally have had.  Likewise for the RQ, RQB, RQC and RQD.

I will fix this.

Meanwhile, once you leave the reservation and try to manually assign addresses you are walking on somewhat shaky ground.  To solve the problem you are seeing all all you need to do is enable devices you want AND disable any devices you don't want BEFORE you manually assign specific addresses to any device.

- Mark

**Comment by agn453** [comment]
Thanks Mark. I appreciate your speedy reply!

It's over 40 years ago that I first had to deal with the intricacies of address and vector assignments on PDP-11/20's and 11/45's - so I understand the problem. As you say, there are certain rules for the first instance of a device and a floating pool of addresses/vectors (that we used to write in pencil on a chart in the System Configuration and Maintenance log!)

Fixing the autoconfig will solve most things - but in some cases it would be nice to be able to manually configure addresses and or the interrupt vectors.  Do you think it would be fairly easy to add the missing SET dev VECTOR=nnn commands for those devices where you could change the jumpers on the boards to select the interrupt vector? They do accept a SET dev ADDRESS=1777nnnnn already and this would allow manual configuration if required.

Tony

**Comment by markpizz** [comment]
The most recent code should fix the problem you were having.

As for working with manually assigned address and vector values...  Well, in theory changes could be made, BUT some would say that given that auto-configure just works, trying to set these things manually would be crazy.  First you've got to manage the 'hardware setup' avoiding conflicts and then you've got to run all over wherever in the OS configuration to manually configure each device to use the manually configured values.

Given the instructions I provided earlier:  
1) Enable all interesting devices 
2) disable all uninteresting devices, 
3) statically set the addresses you want.  
This will result in all devices having assigned addresses and vectors which you can then manually configure the OS to use,  This works.  It works as well as it ever did for all devices which ever could have settable addresses and vectors (the ability to set a device vector has not been removed from any devices).  I'd rather not dig through every device description to add the ability to do things the hard way.

**Comment by agn453** [comment]
Thanks Mark.

I have confirmed that autoconfig is now correctly reassigning addresses/vectors. I built and tested the latest commit on Windows 10 (VS Express 2013 for Desktop and Cygwin 32-bit 2.3.1/gcc 4.9.3) and Mac OS X 10.11.2 (Command Line Tools/Xcode 7.1/Apple LLVM 7.02).

There was just one glitch I had (unrelated to this issue) with expanding the git downloaded simh_master.zip file that I'll raise in another issue.

Your efforts in mentoring SIMH are very much appreciated!

Tony
